docs/hooks.rst

```diff
@@ -1,43 +1,312 @@
-----------------------------------
-The ``preloop`` and ``postloop`` methods run before and after the main loop, respectively.
+There are several pre-existing methods and attributes which you can tweak to
+control the overall behavior of your application before, during, and after the
+command processing loop.
 
-.. automethod:: cmd2.cmd2.Cmd.preloop
+Application Lifecycle Hooks
+---------------------------
+
+You can register methods to be called at the beginning of the command loop::
+
+    class App(cmd2.Cmd):
+        def __init__(self, *args, *kwargs):
+            super().__init__(*args, **kwargs)
+            self.register_preloop_hook(self.myhookmethod)
+
+        def myhookmethod(self):
+            self.poutput("before the loop begins")
+
+To retain backwards compatibility with `cmd.Cmd`, after all registered preloop
+hooks have been called, the ``preloop()`` method is called.
+
+A similar approach allows you to register functions to be called after the
+command loop has finished::
+
+    class App(cmd2.Cmd):
+        def __init__(self, *args, *kwargs):
+            super().__init__(*args, **kwargs)
+            self.register_postloop_hook(self.myhookmethod)
+
+        def myhookmethod(self):
+            self.poutput("before the loop begins")
+
+To retain backwards compatibility with `cmd.Cmd`, after all registered postloop
+hooks have been called, the ``postloop()`` method is called.
+
+Preloop and postloop hook methods are not passed any parameters and any return
+value is ignored.
 
-.. automethod:: cmd2.cmd2.Cmd.postloop
 
 Application Lifecycle Attributes
 --------------------------------
 
-There are numerous attributes (member variables of the ``cmd2.Cmd``) which have a significant effect on the application
-behavior upon entering or during the main loop.  A partial list of some of the more important ones is presented here:
+There are numerous attributes (member variables of the ``cmd2.Cmd``) which have
+a significant effect on the application behavior upon entering or during the
+main loop.  A partial list of some of the more important ones is presented here:
+
+- **intro**: *str* - if provided this serves as the intro banner printed once
+  at start of application, after ``preloop`` runs
+- **allow_cli_args**: *bool* - if True (default), then searches for -t or
+  --test at command line to invoke transcript testing mode instead of a normal
+  main loop and also processes any commands provided as arguments on the
+  command line just prior to entering the main loop
+- **echo**: *bool* - if True, then the command line entered is echoed to the
+  screen (most useful when running scripts)
+- **prompt**: *str* - sets the prompt which is displayed, can be dynamically
+  changed based on application state and/or command results
+
+
+Command Processing Loop
+-----------------------
+
+When you call `.cmdloop()`, the following sequence of events are repeated until
+the application exits:
+
+1. Output the prompt
```

Comment:
Documentation which is similar to this, but now out of date, exists in **integrating.rst** when discussing the behavior of ``onecmd_plus_hooks``.  That file should be updated to be consistent with the new code in this PR.

Reply:
Updated.